joomla/joomla-cms
# Explicitly include default layout paths in JFormField (redux of PR 12960).

Pull Request for Issue # .

### Summary of Changes
This is a re-submission of a previous PR [12960](https://github.com/joomla/joomla-cms/pull/12960). I am now providing the changes suggested in comments on that PR instead of my original changes.

The change here is that `JFormField::getIncludePaths()` now provides the default paths from the `JLayoutFile` rather than providing an empty array and allowing `JLayoutFile` to set the defaults for itself.

The motivation for this is that when creating a new form field type by extending `JFormField` (or one of its subclasses), if you need to override `getIncludePaths()`, then `JLayoutFile` will only take the paths you give it and will no longer load its defaults. This means that your class does not have the option of a fall-back path.

This caused me a problem when extending the subform field type, because the nested fields within the subform also inherit its layout, and thus do not get rendered unless you've created custom layout for them all.

Please see [the original PR](https://github.com/joomla/joomla-cms/pull/12960) for further explanation.

### Testing Instructions

* Create a custom field type overriding subform and implementing its own `getIncludePath()` method. This method should `array_merge()` its custom path with any returned by its parent.
* Also have the class load some fields of any standard field types into the subform.
* Create a custom layout file for the custom subform field type in the path you defined above, but not for any of the standard field types within the subform.
* Without this PR, the fields within the subform will not be rendered.
* After the PR, the fields will be rendered correctly.

### Documentation Changes Required

Documentation for creating custom field types and custom layouts is currently very weak.

The [Custom field types page](https://docs.joomla.org/Creating_a_custom_form_field_type) doesn't even mention being able to use custom layouts. I know this is a niche feature, but it would be good to at least mention it, and include the fact that if you override it you should merge the results with `parent::getIncludePath()` to provide a fallback.
